### ISSUE TITLE
Debugging a weird Concurrent::Delay

### ISSUE THREAD

**Issue by jrochkind** [ruby-concurrency/concurrent-ruby#550]
I'm trying to debug some code that isn't behaving as I expect. It's a somewhat complicated chain of Concurrent::Delay objects that end up triggering each other, so it's hard to reproduce. 

But here's what I'm seeing from a particular delay, and it seems to me to violate the expected contract of Concurrent::Delay and suggest a bug of some kind, but I may be misunderstanding what all these methods do, I'm having trouble finding the right documentation. 

Can you help tell me if this is possibly acceptable behavior for any Concurrent::Delay?

```
   delay.value #=> returns nil immediately
   delay.reason #=> nil
   delay.fulfilled? #=> false
   delay.incomplete? #=> true
```

If the delay is not `fulfilled?`, why did it return from `value` without blocking waiting for it to become fulfilled?  I am not quite sure how I made this happen, but it seems like it must be a bug, Concurrent::Delay shouldn't allow this to happen? or am I misunderstanding the api/contract?

It's somewhat possible I unintentionally triggered a deadlock and/or infinite loop of some kind in my own code. I am trying to figure that out. If I had done that, could that result in the behavior I'm seeing above, without it being a bug?

**Comment by jrochkind** [comment]
Weirder:

    delay.wait   #=> returns self
    delay.pending? #=> true

Huh?  I have no doubt I am doing something impossible in my code, trying to track it down -- I think i'm possibly doing some kind of infinite loop of a delay calling itself (accidentally, via some intermediates).  But no matter what i do, it should not be possible to see what I'm seeing from the Delay, am I right?

**Comment by pitr-ch** [comment]
Hi, thanks for reporting. That definitely does not look right. What implementation of Ruby do you use? I had a quick look into the implementation, but I could not detangle what would be an impact of cyclic delay, @jdantonio could you have a look? @jrochkind you could also try delay from edge, that should be ok (but be aware that the API might change in future, there is ongoing discussion).

**Comment by jrochkind** [comment]
MRI 2.3.1. 

I wasn't able to reduce to an isolated reproduction. I ended up just refactoring my code in a way that no longer seemed to trigger the `nil` value, just kind of bit twiddling to maybe avoid the cycle I'm not entirely sure was there in the first place. It is a bit disconcerting. It is, I'm realizing, hard to figure out the overall picture in a system made up of lots of concurrent parts, one just has to rely on the individual components like a Delay working as per contract.

**Comment by jrochkind** [comment]
@pitr-ch 

Okay, I realized I should try to reproduce with a simple case testing the hypothesis that my problem was an accidentally 'recursive' Delay. Indeed it reproduces that way:

~~~
irb(main):050:0> $delay = Concurrent::Delay.new do
irb(main):051:1*    puts "value: #{$delay.value.inspect}"
irb(main):052:1>    puts "fulfilled?: #{$delay.fulfilled?}"
irb(main):053:1>    puts "complete?: #{$delay.complete?}"
irb(main):054:1>    puts "incomplete?: #{$delay.incomplete?}"
irb(main):055:1>    puts "rejected?: #{$delay.rejected?}"
irb(main):056:1>    puts "reason: #{$delay.reason.inspect}"
irb(main):057:1>    100
irb(main):058:1> end
=> #<Concurrent::Delay:0x007ff2fbc2aa10 @__lock__=#<Mutex:0x007ff2fbc2a998>, @__condition__=#<Thread::ConditionVariable:0x007ff2fbc2a970>, @event=#<Concurrent::Event:0x007ff2fbc2a858 @__lock__=#<Mutex:0x007ff2fbc2a808>, @__condition__=#<Thread::ConditionVariable:0x007ff2fbc2a7e0>, @set=false, @iteration=0>, @dup_on_deref=nil, @freeze_on_deref=nil, @copy_on_deref=nil, @do_nothing_on_deref=true, @executor=nil, @task=#<Proc:0x007ff2fbc2aa38@(irb):50>, @state=:pending, @computing=false>
irb(main):059:0> $delay.value
value: nil
fulfilled?: false
complete?: false
incomplete?: true
rejected?: false
reason: nil
=> 100
~~~

The Delay when accessed inside the Delay thread itself is in a 'contractually unpermitted' state, Delay promises that after calling `value` it will never be incomplete, but violates that commitment in this case. 

The reason this is an issue is becuase it's not hard to _accidentally_ do this, in a much harder to debug manner than the above intentional simple reproduction.  There can be a much more complex backtrace in between the accidental recursion, possibly even involving more threads and more Concurrent entities.  

I had a Delay set up as a lazily-loaded thread-safe global, much like in this example (which is a great use case for Delays, maybe the main use case for delays), and I think what happened is my lazy loading code called a method which called a method which (etc etc) eventually referred to that lazily loaded global itself.  The actual stack trace in my original problem definitely didn't make this obvious though -- it took me a while to brainstorm it even as a possibility. 

I think it's important for Delay not to violate it's contract even in this case -- it should probably instead raise an exception (that is, return a reason) that's some kind of generic deadlock or is actually specific to this case.

**Comment by pitr-ch** [comment]
@jrochkind Thanks you for finding the reproducer! That will help. I agree it should be fixed. Question is how, options I see:
*

**Comment by pitr-ch** [comment]
@jrochkind Thanks you for finding the reproducer! That will help. I agree it should be fixed. Question is how, options I see:
* block on the second value call as well (may block indefinitely, or MRI's deadlock detection will raise)
* detect the cycle and raise error in Delay

**Comment by jrochkind** [comment]
Thanks @pitr-ch . it's clear to me that the best API from the consumer's point of view, is detect and raise immediately.  A self-referencing cycle in a Delay is never going to work, it's impossible -- but easy to do accidentally (I can provide more realistic examples if this isn't clear how easy it is to do accidentally in a hard to debug way) -- it should fail quick with a clear error.  Blocking indefinitely would make it harder, not easier to debug. 

As long as it fails immediately with a raise (that would probably be revealed as a Delay `reason` because of the nature of where this occurs), I think it's acceptable -- a custom error saying exactly what the problem is would be ideal, but a generic MRI deadlock detection error would also be fine, if it could be guaranteed to happen. 

What I have no idea about is how hard it is to do the cycle detection to make it raise.

**Comment by jrochkind** [comment]
Ideal would of course be if the exception included a backtrace which actually included the cycle (Delay.value twice in it). I don't know if a generic MRI deadlock exception would, probably depends on where/how it's raised.

**Comment by jrochkind** [comment]
(PS: And of course we're talking about "MRI" deadlock detection error, but this needs to do something reasonable -- and ideally consistent -- on JRuby etc too)

**Comment by jdantonio** [comment]
@jrochkind Good catch. Thank you so much for all the work you went to in debugging this! Honestly, a recursive Delay is a situation I never gave any thought to. I agree with you that an immediate exception should be the result. I'll work on it this weekend. I'm not sure how it'll work but I'll see what I can do.

**Comment by pitr-ch** [comment]
This got left behind. The delay will be deprecated, so closing in favour of new implementation which will be merged in core with 1.1 release. New implementation deadlocks (no inconsistent state), however using value is discouraged and it is not possible to build a cycle with preferred chaining methods. Therefore I won't investigate this issue further for now.

**Comment by jrochkind** [comment]
@pitr-ch Hey, 15 months later.... any chance of fixing Delay to make it reliable?
